### PR TITLE
tracing: unify prepared statement info into a single struct

### PR DESCRIPTION
**Pull request by psarna**
The tracing code assumes that query_option_names and query_option_values
vectors always have the same length as the prepared_statements vector,
but it's not true. E.g. if one of the statements in a batch is
incorrect, it will create a discrepancy between the number of prepared
statements and the number of bound names and values, which currently
leads to a segmentation fault. It's better to simply skip generating
this additional information, because it's incomplete.

Tested manually with code that triggers a failure while executing
a batch statement, because the Python driver performs driver-side
validation and thus it's hard to create a test case which triggers
the problem.

Fixes https://github.com/scylladb/scylla/issues/9220

### CODE REVIEW

**Review by vladzcloudius — CHANGES_REQUESTED** [review]
I don't like this patch. It works around a problem instead of fixing the root cause, which is a missing validation on a cql_server level that is supposed to validate exactly what you are trying to work around. 

IMO the assumption you mentioned is absolutely valid at a Tracing level and it has to be validated at higher layers before the corresponding Tracing function is called.

**Comment by psarna** [comment]
v2:
 - changed throwing to checking if sizes match and skipping the param map generation if they don't

**Comment by vladzcloudius** [comment]
While v2 is better than v1 but it's still a workaround and not a fix.

Please, see a detailed description of this regression together with a proposal for a fix here: https://github.com/scylladb/scylla/issues/9220#issuecomment-902758002

@avikivity @psarna FYI

**Comment by psarna** [comment]
v3:
 - unified all 3 vectors to a single vector, which renders the size mismatch impossible

File: tracing/trace_state.cc [hunk 1]
```diff
@@ -208,11 +210,12 @@ void trace_state::build_parameters_map() {
         // queries CQL command, where X is an index of the parameter in a corresponding query and Y is an index of the
         // corresponding query in the BATCH.
         if (prepared_statements.size() == 1) {
-            build_parameters_map_for_one_prepared(prepared_statements[0], vals.query_option_names[0], vals.query_option_values[0], "param");
+            build_parameters_map_for_one_prepared(prepared_statements[0].statement, prepared_statements[0].query_option_names, prepared_statements[0].query_option_values, "param");
```
Comment (vladzcloudius) [inline tracing/trace_state.cc:213]: It's a bit of a nitpick but why not to define `auto& stmt_info` like you did for a BATCH below to shorten the `build_parameters_map_for_one_prepared()` call string and avoid typos in the future?

**Review by vladzcloudius — CHANGES_REQUESTED** [review]
Looks good in general.
Have a few nitpicks though.

File: tracing/trace_state.cc (hunk 1 above)
Comment (psarna) [reply tracing/trace_state.cc:213]: no reason - I'll use a helper variable in v2

File: tracing/trace_state.cc [hunk 2]
```diff
@@ -144,20 +148,18 @@ void trace_state::set_user_timestamp(api::timestamp_type val) {
 }
 
 void trace_state::add_prepared_statement(prepared_checked_weak_ptr& prepared) {
-    _params_ptr->prepared_statements.emplace_back(prepared->checked_weak_from_this());
+    _params_ptr->prepared_statements.push_back(params_values::prepared_statement_info{prepared->checked_weak_from_this(), {}, {}});
```
[the remark this answers is not in the document]
Comment (psarna) [reply tracing/trace_state.cc:151]: I started with `emplace_back`, but it failed to deduce the types from `{}` passed as second and third parameters, and the definitions of these types are really long and I'd rather not explicitly copy them. So, I decided to push_back, since it's likely going to be compiled to identical code anyway.

[the remark this answers is not in the document]
Comment (vladzcloudius) [reply tracing/trace_state.cc:151]: I see. Could you, please, add an inline comment above this line then so that we would understand why it's this way in the future. ;)

[the remark this answers is not in the document]
Comment (psarna) [reply tracing/trace_state.cc:151]: I don't follow what would be the contents of such a comment, or its purpose. In my opinion, comments should describe non-obvious semantics, which is not the case here - it's clear that we're just putting stuff into a vector. Both emplace_back and push_back makes perfect sense here, it's just that the code will get longer in case of emplace_back due to having to explicitly state all the types. It's a kind of a question that can be answered via review, and doesn't warrant a comment in the code. We already have multiple precedents in the code similar to this one, none of them are commented:
```
[sarna@localhost scylla]$ git grep push_back.*\,\ \{\}
clustering_ranges_walker.hh:                    rts.push_back(range_tombstone_change(_current_end, {}));
raft/server.cc:    _reads.push_back({rid->first, rid->second, {}});
repair/row_level.cc:                            row_list.push_back(repair_row({}, {}, dk_ptr, {}, is_dirty_on_master::no, std::move(mf)));
sstables/mutation_fragment_filter.hh:                pr.rts.push_back(range_tombstone_change(_fwd_end, {}));
utils/logalloc.cc:        _backtraces.push_back(std::make_unique<backtrace_entry>(backtrace_entry{current_backtrace(), {}}));
```

[the remark this answers is not in the document]
Comment (vladzcloudius) [reply tracing/trace_state.cc:151]: Well, using `emplace` makes more sense here since you are constructing the item in-place.
However you DO have a good-enough reason to use push_back which is not obvious unless you explain it.

So, the comment could be something like that: 
```
// Use push_back() instead of emplace_back() to avoid long explicit casts
```

File: tracing/trace_state.cc [hunk 3]
```diff
@@ -55,15 +55,19 @@ namespace tracing {
 logging::logger trace_state_logger("trace_state");
 
 struct trace_state::params_values {
+    struct prepared_statement_info {
+        prepared_checked_weak_ptr statement;
+        std::optional<std::vector<sstring_view>> query_option_names;
+        std::vector<cql3::raw_value_view> query_option_values;
```
Comment (vladzcloudius) [inline tracing/trace_state.cc:61]: You probably want to introduce a constructor that receives `prepared_checked_weak_ptr`. See more below.

File: tracing/trace_state.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (vladzcloudius) [reply tracing/trace_state.cc:151]: I think the initialization is somewhat bogus: you initialize an `optional<vector>` with an empty vector which means that until it's reassigned with an actual value it would be holding an actual value (empty vector) which is both logically confusing (it's supposed to be uninitialized) but also inefficient because that empty vector value is not needed when unnamed prepared parameters semantics is used (most common one) and then this vector value is going to be discarded the moment this field is going to be assigned `nullopt`.

Long story short, it would be better if you would implement that constructor I mentioned above. This would also allow you to use `emplace_back()` here and keep the code line both short and self-explanatory.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: tracing/trace_state.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (psarna) [reply tracing/trace_state.cc:151]: It's not bogus - intializing an `optional<T>` with `{}` is equivalent to initializing it with a disengaged optional, just like passing `std::nullopt`. No empty vector would be constructed and there's no inefficiency. Regardless, I don't mind creating a custom constructor if it makes the review process go forward, I'll push an update next week.
